File: libsmb/src/main/java/com/xyoye/libsmb/SmbManager.java
package com.xyoye.libsmb;

import com.xyoye.libsmb.controller.Controller;
import com.xyoye.libsmb.controller.JCIFSController;
import com.xyoye.libsmb.controller.JCIFS_NGController;
import com.xyoye.libsmb.controller.SMBJController;
import com.xyoye.libsmb.controller.SMBJ_RPCController;
import com.xyoye.libsmb.exception.SmbLinkException;
import com.xyoye.libsmb.info.SmbLinkInfo;
import com.xyoye.libsmb.info.SmbType;
import com.xyoye.libsmb.utils.SmbUtils;

/**
 * Created by xyoye on 2019/12/20.
 */

public class SmbManager {
    private SmbType mSmbType;
    private boolean isLinked;
    private Controller controller;
    private SmbLinkException smbLinkException;

    private boolean smbJRPCEnable = true;
    private boolean smbJEnable = true;
    private boolean jcifsNGEnable = true;
    private boolean jcifsEnable = true;

    private static class Holder {
        static SmbManager instance = new SmbManager();
    }

    private SmbManager() {
        smbLinkException = new SmbLinkException();
    }

    public static SmbManager getInstance() {
        return Holder.instance;
    }

    /**
     * link to the smb server from smbV2 to smbV1
     *
     * @param smbLinkInfo link data
     */
    public boolean linkStart(SmbLinkInfo smbLinkInfo) {

        smbLinkException.clearException();

        if (!smbLinkInfo.isAnonymous()) {
            if (SmbUtils.containsEmptyText(smbLinkInfo.getAccount(), smbLinkInfo.getAccount())) {
                throw new NullPointerException("Account And Password Must NotNull");
            }
        }

        //SMB V2
        isLinked = true;

        if (jcifsNGEnable) {
            controller = new JCIFS_NGController();
            if (controller.linkStart(smbLinkInfo, smbLinkException)) {
                mSmbType = SmbType.JCIFS_NG;
                return true;
            }
        }

        if (smbJRPCEnable) {
            controller = new SMBJ_RPCController();
            if (controller.linkStart(smbLinkInfo, smbLinkException)) {
                mSmbType = SmbType.SMBJ_RPC;
                return true;
            }
        }

        if (smbJEnable && !SmbUtils.isTextEmpty(smbLinkInfo.getRootFolder())) {
            controller = new SMBJController();
            if (controller.linkStart(smbLinkInfo, smbLinkException)) {
                mSmbType = SmbType.SMBJ;
                return true;
            }
        }

        //SMB V1
        if (jcifsEnable) {
            controller = new JCIFSController();
            if (controller.linkStart(smbLinkInfo, smbLinkException)) {
                mSmbType = SmbType.JCIFS;
                return true;
            }
        }

        isLinked = false;
        return false;
    }

    /**
     * get smb tools type
     */
    public String getSmbType() {
        return mSmbType == null ? "" : SmbType.getTypeName(mSmbType);
    }

    /**
     * is the link successful
     */
    public boolean isLinked() {
        return isLinked;
    }

    /**
     * get link controller
     */
    public Controller getController() {
        return controller;
    }

    /**
     * link error info
     */
    public SmbLinkException getException() {
        return smbLinkException;
    }

    public void setEnable(boolean jcifsNGEnable, boolean smbJRPCEnable, boolean smbJEnable, boolean jcifsEnable) {
        this.jcifsNGEnable = jcifsNGEnable;
        this.smbJRPCEnable = smbJRPCEnable;
        this.smbJEnable = smbJEnable;
        this.jcifsEnable = jcifsEnable;
    }
}

File: libsmb_kotlin/src/main/java/com/xyoye/libsmb_kotlin/SmbManager.kt
package com.xyoye.libsmb_kotlin

import com.xyoye.libsmb_kotlin.controller.*
import com.xyoye.libsmb_kotlin.exception.SmbLinkException
import com.xyoye.libsmb_kotlin.info.SmbLinkInfo
import com.xyoye.libsmb_kotlin.info.SmbType

/**
 * Created by xyoye on 2019/12/25.
 */

class SmbManager private constructor() {

    companion object {
        fun getInstance(): SmbManager {
            return Holder.instance
        }
    }

    private object Holder {
        val instance = SmbManager()
    }

    var isLinked = false
    var controller: Controller? = null
    var smbLinkException = SmbLinkException()

    var smbJRPCEnable = false
    var smbJEnable = false
    var jcifsNgEnable = false
    var jcifsEnable = false

    private var mSmbType: SmbType? = null
    private var linkCallback: LinkCallback? = null

    fun linkStart(smbLinkInfo: SmbLinkInfo) {
        smbLinkException.clearException()

        if (!smbLinkInfo.isAnonymous) {
            if (smbLinkInfo.account.isEmpty() || smbLinkInfo.password.isEmpty()) {
                throw NullPointerException("Account And Password Must Not Null")
            }
        }

        isLinked = true
        if (smbJRPCEnable) {
            linkCallback?.onLinkChange("SMBJ_RPC")
            controller = SMBJ_RPCController()
            if (controller!!.linkStart(smbLinkInfo, smbLinkException)) {
                mSmbType = SmbType.SMBJ_RPC
                linkCallback?.onSuccess()
                return
            }
        }

        if (jcifsNgEnable) {
            linkCallback?.onLinkChange("JCIFS_NG")
            controller = JCIFS_NGController()
            if (controller!!.linkStart(smbLinkInfo, smbLinkException)) {
                mSmbType = SmbType.JCIFS_NG
                linkCallback?.onSuccess()
                return
            }
        }

        if (smbJEnable && smbLinkInfo.rootFolder.isNotEmpty()) {
            linkCallback?.onLinkChange("SMBJ")
            controller = SMBJController()
            if (controller!!.linkStart(smbLinkInfo, smbLinkException)) {
                mSmbType = SmbType.SMBJ
                linkCallback?.onSuccess()
                return
            }
        }

        if (jcifsEnable) {
            linkCallback?.onLinkChange("JCIFS")
            controller = JCIFSController()
            if (controller!!.linkStart(smbLinkInfo, smbLinkException)) {
                mSmbType = SmbType.JCIFS
                linkCallback?.onSuccess()
                return
            }
        }

        isLinked = false
        linkCallback?.onFailed()
    }

    fun getSmbType(): String {
        return if (mSmbType == null) "" else mSmbType!!.name
    }

    fun setEnable(smbJRPCEnable: Boolean, smbJEnable: Boolean, jcifsNgEnable:Boolean, jcifsEnable: Boolean){
        this.smbJRPCEnable = smbJRPCEnable
        this.smbJEnable = smbJEnable
        this.jcifsNgEnable = jcifsNgEnable
        this.jcifsEnable = jcifsEnable
    }

    fun setLinkCallback(linkCallback: LinkCallback){
        this.linkCallback = linkCallback
    }

    interface LinkCallback {
        fun onLinkChange(type: String)

        fun onSuccess()

        fun onFailed()
    }
}